src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -413,6 +477,71 @@ public static UIntPtr CompareExchange(ref UIntPtr location1, UIntPtr value, UInt
             return (UIntPtr)CompareExchange(ref Unsafe.As<UIntPtr, int>(ref location1), (int)value, (int)comparand);
 #endif
         }
+
+        /// <summary>Compares two instances of the specified type <typeparamref name="T"/> for equality and, if they are equal, replaces the first one.</summary>
+        /// <param name="location1">The destination, whose value is compared with <paramref name="comparand"/> and possibly replaced.</param>
+        /// <param name="value">The value that replaces the destination value if the comparison results in equality.</param>
+        /// <param name="comparand">The object that is compared to the value at <paramref name="location1"/>.</param>
+        /// <returns>The original value in <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of <paramref name="location1"/> is a null pointer.</exception>
+        /// <exception cref="NotSupportedException">An unsupported <typeparamref name="T"/> is specified.</exception>
+        /// <typeparam name="T">
+        /// The type to be used for <paramref name="location1"/>, <paramref name="value"/>, and <paramref name="comparand"/>.
+        /// This type must be a reference type, an enum type (i.e. typeof(T).IsEnum is true), or a primitive type (i.e. typeof(T).IsPrimitive is true).
+        /// </typeparam>
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [return: NotNullIfNotNull(nameof(location1))]
+        public static unsafe T CompareExchange<T>(ref T location1, T value, T comparand)
+        {
+            // Handle all reference types with CompareExchange(ref object, ...).
+            if (!typeof(T).IsValueType)
+            {
+                object? result = CompareExchange(ref Unsafe.As<T, object?>(ref location1), value, comparand);
+                return Unsafe.As<object?, T>(ref result);
+            }
+
+            // Handle everything else with a CompareExchange overload for the unsigned integral type of the corresponding size.
+            // Only primitive types and enum types (which are backed by primitive types) are supported.
+            if (!typeof(T).IsPrimitive && !typeof(T).IsEnum)
```

Comment:
Are we sure we want to accept floating point data here? Since the code does bitwise equality, having them here could be misleading.

Reply:
> Are we sure we want to accept floating point data here? Since the code does bitwise equality, having them here could be misleading.

How is that different from the existing overloads strongly-typed to take float/double?
https://github.com/dotnet/runtime/blob/ac0eeb015a0266adb06696bb58c21e5bfe6d9b56/src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs#L366-L378

Reply:
I would find it incredibly surprising if the behavior of `Interlocked.CompareExchange(ref float, ...)` and `Interlocked.CompareExchange<float>(ref float, ...)` differed. If we're talking about changing the behavior of both, sure, that's a breaking change that could be considered, but as you say it's not related to this PR.